File: section-03-basic-maths/09-check-for-perfect-number/check-for-perfect-number-opt.cs
using System;

public class Solution {
    /* Function to find whether the
    number is perfect or not */
    public bool IsPerfect(int n) {
        // Edge case
        if (n == 1)
            return false;
            
        /* Variable to store the sum
        of all proper divisors */
        long sum = 0;
        
        // Loop from 1 to square root of n
        for (int i = 1; (long)i * i <= n; ++i) {
            
            // Check if i is a proper divisor
            if (n % i == 0) {
                // Update sum
                sum = sum + i;
                
                /* Add the counterpart divisor
                if it's different from i and
                if it is not n itself */
                if (i != 1 && (long)i * i != n) {
                    sum = sum + (n / i);
                }
            }
        }
        
        // Compare sum and n
        if (sum == n) return true;
        return false;
    }
}

public class Program {
    public static void Main(string[] args) {
        int n = 6;

        /* Creating an instance of 
        Solution class */
        Solution sol = new Solution();

        /* Function call to find whether the
        given number is perfect or not */
        bool ans = sol.IsPerfect(n);

        if (ans) {
            Console.WriteLine($"{n} is a perfect number.");
        } else {
            Console.WriteLine($"{n} is not a perfect number.");
        }
    }
}

File: section-03-basic-maths/09-check-for-perfect-number/check-for-perfect-number-opt.py
import math
class Solution:
    def isPerfect(self, n: int) -> bool:
        if n==1:
            return False
        sum=0
        for i in range(1,int(math.sqrt(n))+1):
            if(n %i ==0):
                sum += i
                if i != 1 and i !=n//i:
                    sum+=n//i
        
        if(sum==n):
            return True
        return False


""" 
1. When `i` divides `n`, there are always two divisors: `i` and `n//i`.
2. We already add `i`, so this condition decides whether to also add its pair `n//i`.
3. `i != 1` ensures we don’t add `n` itself (since when `i = 1`, `n//i = n`, which is not a proper divisor).
4. `i != n//i` ensures we don’t count the same number twice when `n` is a perfect square.
5. Example: for `n = 36` and `i = 6`, `n//i = 6`, so we add `6` only once (not twice), and for `i = 1`, we add only `1` and skip `36`.


"""